### PR TITLE
Make metrics serializable

### LINKED ISSUE #2999: Add `state_dict` and `load_state_dict` to metrics
## 🚀 Feature

To add `state_dict` and `load_state_dict` to metrics to be able to save them to checkpoints and update them later.

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Fixes #2999

### CODE REVIEW

**Comment by sadra-barikbin** [comment]
Shall I add `_state_dict_all_req_keys` attribute to all of metrics?
Shall I add tests for `state_dict` and `load_state_dict` for each metric?

File: ignite/metrics/metric.py [hunk 1]
```diff
@@ -546,6 +550,41 @@ def is_attached(self, engine: Engine, usage: Union[str, MetricUsage] = EpochWise
...
+        for attr_name in self._state_dict_all_req_keys:
+            attr = getattr(self, attr_name)
+            if idist.get_world_size() == 1:
+                state[attr_name] = [attr]
+            else:
+                if isinstance(attr, (float, torch.Tensor)):
+                    state[attr_name] = cast(List[Any], idist.all_gather(attr))
+                else:
+                    state[attr_name] = [None] * idist.get_world_size()
+                    torch_dist.all_gather_object(state[attr_name], attr)
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:570]: Let's raise an error here now. We can add a support for picklable objects later with `idist.all_gather`.

File: ignite/metrics/metric.py [hunk 2]
```diff
@@ -546,6 +550,41 @@ def is_attached(self, engine: Engine, usage: Union[str, MetricUsage] = EpochWise
...
+    def state_dict(self) -> OrderedDict:
+        """Method returns state dict with attributes of the metric specified in its
+        `_state_dict_all_req_keys` attribute. Can be used to save internal state of the class.
+
+        If there's an active distributed configuration, some collective operations is done and
+        the list of values across ranks is saved under each attribute's name in the dict.
+        """
+        state = OrderedDict()
+        for attr_name in self._state_dict_all_req_keys:
+            attr = getattr(self, attr_name)
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:562]: Let's add a check if attr_name does not exist.

File: ignite/metrics/metric.py [hunk 3]
```diff
@@ -546,6 +550,41 @@ def is_attached(self, engine: Engine, usage: Union[str, MetricUsage] = EpochWise
...
+
+        If there's an active distributed configuration, the process uses its rank to pick the proper value from
+        the list of values saved under each attribute's name in the dict.
+
+        Args:
+            state_dict: a dict containing attributes of the metric specified in its `_state_dict_all_req_keys`
+                attribute.
+        """
+        super().load_state_dict(state_dict)
+        rank = idist.get_local_rank()
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:584]: It should be rank and not local_rank here, right ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @sadra-barikbin !
The approach looks OK to me, let's update all metrics.

File: docs/source/engine.rst [hunk 4]
```diff
@@ -82,8 +82,9 @@ store the trainer and then resume the training. For example:
     optimizer = ...
     lr_scheduler = ...
     data_loader = ...
+    metric = ...
 
-    to_save = {'trainer': trainer, 'model': model, 'optimizer': optimizer, 'lr_scheduler': lr_scheduler}
+    to_save = {'trainer': trainer, 'model': model, 'optimizer': optimizer, 'lr_scheduler': lr_scheduler, 'metric': metric}
```
Comment (vfdev-5) [inline docs/source/engine.rst:87]: Ideally, we would like to get that in a more automatic way. Also if we want to save multiple metrics adding them one by one to save/load can be a pain.

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @sadra-barikbin

File: ignite/metrics/running_average.py [hunk 5]
```diff
@@ -87,6 +87,9 @@ def log_running_avg_metrics():
     """
 
     required_output_keys = None
+    # TODO Shall we put `src` here? Then we should add a new branch for metric-typed attributes in `state_dict`
+    # and `load_state_dict`. Examples; This class; `Rouge` which has a `List[_BaseRouge]`.
+    _state_dict_all_req_keys = ("_value",)
```
Comment (sadra-barikbin) [inline ignite/metrics/running_average.py:92]: @vfdev-5 , What's your thoughts on this?

Comment (vfdev-5) [reply ignite/metrics/running_average.py:92]: Let's add `src` and in state_dict/load_state_dict put an explicit check for `isinstance(attr, Metric)` and call `state[attr_name].update(attr.state_dict)` or something like that.
Do you have any other options ?

File: docs/source/engine.rst (hunk 4 above)
Comment (sadra-barikbin) [reply docs/source/engine.rst:87]: Could you please reiterate what you mean? If we have multiple metrics, is it cumbersome to do:
```python
to_save = {..., 'metric1': metric1, 'metric2': metric2, 'metric3': metric3}
```
?

Comment (vfdev-5) [reply docs/source/engine.rst:87]: Yes, it is too verbose, let's try to find a way to simplify that.
